File: easy/reverse-linked-list/recursive.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        rev(head, None)
    }
}

fn rev(curr: Option<Box<ListNode>>, prev: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    match curr {
        None => prev,
        Some(mut curr) => {
            let next = curr.next;
            curr.next = prev;
            rev(next, Some(curr))
        }
    }
}

File: easy/reverse-linked-list/recursive.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* recur(struct ListNode* curr, struct ListNode* prev) {
    if (!curr)
        return prev;
    struct ListNode *next = curr->next;
    curr->next = prev;
    return recur(next, curr);
}

struct ListNode* reverseList(struct ListNode* head) {
    return recur(head, NULL);
}
